operate/utils/__init__.py

```diff
@@ -23,6 +23,21 @@
 import time
 import typing as t
 from pathlib import Path
+from threading import Lock
+
+
+class SingletonMeta(type):
+    """A metaclass for creating thread-safe singleton classes."""
+
+    _instances: t.Dict[t.Type, t.Any] = {}
+    _lock: Lock = Lock()
+
+    def __call__(cls, *args: t.Any, **kwargs: t.Any) -> t.Any:
+        """Override the __call__ method to control instance creation."""
+        with cls._lock:
+            if cls not in cls._instances:
+                cls._instances[cls] = super().__call__(*args, **kwargs)
+        return cls._instances[cls]
```

Comment:
This seems thread safe at creation time, but is this also trhead-safe at call time (i.e., when the class is being used)?